extensions/amp-a4a/0.1/amp-a4a.js

```diff
@@ -0,0 +1,706 @@
+   * TODO(keithwrightbos@): report error cases
+   */
+  getAmpAdMetadata_(creative) {
+    window.top.creative = creative;
+    const metadataStart = creative.lastIndexOf(METADATA_STRING);
+    const metadataEnd = creative.lastIndexOf('</script>');
+    if (metadataStart < 0 || metadataEnd < 0) {
+      // Couldn't find a metadata blob.
+      return null;
+    }
+    try {
+      return this.buildCreativeMetaData_(JSON.parse(
+        creative.slice(metadataStart + METADATA_STRING.length, metadataEnd)));
+    } catch (err) {
+      return null;
+    }
+  }
+
+  /**
+   * @param {!Object} JSON extraced from creative
+   * @return {!CreativeMetaDataDef} if valid, null otherwise
+   * @private
+   */
+  buildCreativeMetaData_(metaDataObj) {
+    const metaData = {};
+    const isValidOffsetArray = function(ary) {
+      return isArray(ary) && ary.length == 2 &&
+          typeof ary[0] === 'number' &&
+          typeof ary[1] === 'number';
+    };
+    metaData.bodyUtf16CharOffsets = metaDataObj['bodyUtf16CharOffsets'];
+    if (!isValidOffsetArray(metaData.bodyUtf16CharOffsets)) {
+      // Invalid/Missing body offsets array.
+      throw new Error('Invalid/missing body offsets');
+    }
+    if (metaDataObj['cssUtf16CharOffsets']) {
+      metaData.cssUtf16CharOffsets = metaDataObj['cssUtf16CharOffsets'];
+      if (!isValidOffsetArray(metaData.cssUtf16CharOffsets)) {
+        throw new Error('Invalid CSS offsets');
+      }
+    }
+    // Validate array of two member number arrays
+    if (metaDataObj['cssReplacementRanges']) {
+      metaData.cssReplacementRanges = metaDataObj['cssReplacementRanges'];
+      if (!isArray(metaData.cssReplacementRanges)) {
+        throw new Error('Invalid CSS replacement ranges');
+      }
+      for (let i = 0; i < metaData.cssReplacementRanges.length; i++) {
+        if (!isValidOffsetArray(metaData.cssReplacementRanges[i])) {
+          throw new Error('Invalid CSS replacement ranges');
+        }
+      }
+    }
+    if (metaDataObj['bodyAttributes']) {
+      metaData.bodyAttributes = metaDataObj['bodyAttributes'];
+      if (typeof metaData.bodyAttributes !== 'string') {
+        throw new Error('Invalid body attributes');
+      }
+    }
+    if (metaDataObj['customElementExtensions']) {
+      metaData.customElementExtensions = metaDataObj['customElementExtensions'];
+      if (!isArray(metaData.customElementExtensions)) {
+        throw new Error('Invalid extensions');
+      }
+    }
+    if (metaDataObj['customStylesheets']) {
+      // Expect array of objects with at least one key being 'href' whose value
+      // is URL.
+      metaData.customStylesheets = metaDataObj['customStylesheets'];
+      if (!isArray(metaData.customStylesheets)) {
+        throw new Error('Invalid custom stylesheets');
+      }
+      for (let i = 0; i < metaData.customStylesheets.length; i++) {
+        const stylesheet = metaData.customStylesheets[i];
+        if (!isObject(stylesheet) || !stylesheet['href'] ||
```

Comment:
So A4A is allowed to include external stylesheets? Doesn't that make it invalid AMP?